src/functions/leastSquaresDirect.jl

```diff
@@ -45,8 +45,17 @@ function LeastSquaresDirect(A::M, b::V, lambda::R) where {R <: Real, C <: Union{
     LeastSquaresDirect{R, C, M, V, SuiteSparse.CHOLMOD.Factor{C}}(A, b, lambda)
 end
 
+# Adjoint/Transpose versions
+function LeastSquaresDirect(A::M, b::V, lambda::R) where {R <: Real, C <: Union{R, Complex{R}}, M <: TransposeOrAdjoint{<:DenseMatrix{C}}, V <: AbstractVector{C}}
+    LeastSquaresDirect(copy(A), b, lambda)
+end
+
+function LeastSquaresDirect(A::M, b::V, lambda::R) where {R <: Real, C <: Union{R, Complex{R}}, I <: Integer, M <: TransposeOrAdjoint{<:SparseMatrixCSC{C, I}}, V <: AbstractVector{C}}
+    LeastSquaresDirect(copy(A), b, lambda)
```

Comment:
Why the need to `copy` here (similarly in line 50)? If I’m not wrong, `adjoint(adjoint(X)) === X` for any `X` (same for `transpose`), see

https://github.com/JuliaLang/julia/blob/38b38cfdd5b1a5bdf1a613a6180d2aaadf06a632/stdlib/LinearAlgebra/src/adjtrans.jl#L152-L153

so maybe we can use `adjoint(A)` here and avoid further allocations? Of course, that would require that the `Transpose` and `Adjoint` cases are coded separately.

Reply:
Actually, one could use `A.parent` and catch both `Transpose` and `Adjoint` with one implementation, but that would rely on the internals of some structure owned by others, which seems weird to do. I’m not sure 🤔

Reply:
The function is catching the case
`LeastSquares(A', b, lambda)` or `LeastSquares(transpose(A), b, lambda)`
i.e where we want to work with the adjoint of some allocated matrix.
The way I see it is that we have three options if we dont want to copy:

* We save `A`, and have a flag for `adjoint/transpose/identity` and then have separate implementations based on that flag for `prox!`, `factor_step!`, `solve_step!`, `gradient!`.
* We save A' in the struct, which means that A::Adjoint{Float64,Matrix{Float64}}, for example. But then we no longer have that `S = A'*A`, is of type `M` so we need a new type parameter, say M2. We want to be able to infer the factorization type `F` in this case, which for the ones we already inffer I guess is `Cholesky{C,M2}`, and `SuiteSparse.CHOLMOD.Factor{C}` anyway.
* We only fix `warn()` -> `@warn`. which will result in a LeastSquares with `F=Factorization`. But it doesn't even look like `prox!` will work in this case, I cant see a version of `solve_step!` that would catch it. It would probably be possible to write a type-unstable constructor instead that creates a LeastSquares with the proper `F`, but again, this requires a bit of re-coding.

The second one I think would be all right. But in either case, the constructor allocates both `S = A'*A` or `S = A*A'` as well as a factorization and a matrix in the first `prox!`: `cholesky(f.S + I/lamgam)`.
To be fair, if `A` is really tall or fat, these allocations will "only" be `min(m,n)^2` compared to a `copy(A)` which is `m*n`, but I'm not sure that one could argue that `copy(A)` is really that much worse than what we are already doing. The best argument I can find is that if the user only wants to compute gradients, then the constructor could have been allocation free (except for a vector), but that would require a larger rewrite for the case we already support.

Reply:
Disregard my comment, I was wrong. It’s fine to use `copy`, since that’s definitely correct.